File: 2015/07/rust/src/main.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{ BufRead, BufReader };

// --- Constants ---
const INPUT_FILE:&'static str = "../input/input.txt";

fn solve(op: &str, wires: &HashMap<String, String>, cache: &mut HashMap<String, u16>) -> u16 {
    if let Ok(n) = op.trim().parse::<u16>() {
        return n;
    }
    if let Some(n) = cache.get(op) {
        return *n;
    }
    let ls:Vec<&str> = wires.get(op).unwrap().split(" ").collect();
    let r = if ls.contains(&"AND") {
        solve(ls[0], wires, cache) & solve(ls[2], wires, cache)
    } else if ls.contains(&"OR") {
        solve(ls[0], wires, cache) | solve(ls[2], wires, cache)
    } else if ls.contains(&"LSHIFT") {
        solve(ls[0], wires, cache) << solve(ls[2], wires, cache)
    } else if ls.contains(&"RSHIFT") {
        solve(ls[0], wires, cache) >> solve(ls[2], wires, cache)
    } else if ls.contains(&"NOT") {
        !solve(ls[1], wires, cache)
    } else {
        solve(ls[0], wires, cache)
    };
    cache.insert(op.to_string(), r);
    return r;
}

// --- Main function ---
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut wires: HashMap<String, String> = HashMap::new();
    let mut cache: HashMap<String, u16> = HashMap::new();

    // --- Read and parse the input file ---
    let file = File::open(INPUT_FILE)?;
    for line in BufReader::new(file).lines() {
        let line = line.unwrap();
        let v: Vec<&str> = line.split(" -> ").collect();
        wires.insert(v[1].trim().to_string(), v[0].trim().to_string());
    }

    // --- Puzzle 1 ---
    let r = solve("a", &wires, &mut cache);
    println!("1. Wire a: {}", r);

    // --- Puzzle 2 ---
    wires.insert("b".to_string(), format!("{}", r));
    cache.clear();
    let r = solve("a", &wires, &mut cache);
    println!("2. Wire a: {}", r);

    Ok(())
}

File: 2015/07/javascript/main.mjs
import fs from "node:fs";
import readline from "node:readline";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";

function solve(op, wires, cache) {
	if (!isNaN(op)) return Number(op);
	if (cache[op] !== undefined) return cache[op];

	function _solve(o) { return solve(o, wires, cache); }

	let r;
	const ls = wires[op].split(" ");
	if (ls.includes("AND"))         r = _solve(ls[0]) & _solve(ls[2]);
	else if (ls.includes("OR"))     r = _solve(ls[0]) | _solve(ls[2]);
	else if (ls.includes("LSHIFT")) r = _solve(ls[0]) << _solve(ls[2]);
	else if (ls.includes("RSHIFT")) r = _solve(ls[0]) >> _solve(ls[2]);
	else if (ls.includes("NOT"))    r = ~ _solve(ls[1]);
	else r = _solve(ls[0]);
	cache[op] = r;
	return r;
}

// --- Variables ---
let wires = {};
let cache = {};

// --- Read and parse the input file ---
const stream = fs.createReadStream(INPUT_FILE);
const reader = readline.createInterface({ input: stream, crlDelay: Infinity });
reader.on("line", line => {
	const ls = line.split(" -> ");
	wires[ls[1].trim()] = ls[0].trim();
});
await new Promise(res => reader.once("close", res));

// --- Puzzle 1 ---
let r = solve("a", wires, cache);
console.log("1. Wire a: ", r.toLocaleString());

// --- Puzzle 2 ---
wires["b"] = r.toString();
cache = {};
r = solve("a", wires, cache);
console.log("2. Wire a: ", r.toLocaleString());
